File: src/main/java/br/com/leonardoferreira/jirareport/config/security/TokenAuthenticationProvider.java
package br.com.leonardoferreira.jirareport.config.security;

import br.com.leonardoferreira.jirareport.domain.vo.Account;
import br.com.leonardoferreira.jirareport.service.TokenService;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.AuthenticationProvider;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.jwt.Jwt;
import org.springframework.security.jwt.JwtHelper;
import org.springframework.security.jwt.crypto.sign.RsaVerifier;
import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken;
import org.springframework.security.web.authentication.preauth.PreAuthenticatedCredentialsNotFoundException;
import org.springframework.stereotype.Component;

@Component
public class TokenAuthenticationProvider implements AuthenticationProvider {

    @Autowired
    private TokenService tokenService;

    @Override
    public Authentication authenticate(final Authentication authentication) throws AuthenticationException {
        try {
            Account account = tokenService.decode((String) authentication.getPrincipal());
            return new PreAuthenticatedAuthenticationToken(account, account.getPassword(), account.getAuthorities());
        } catch (Exception e) {
            throw new PreAuthenticatedCredentialsNotFoundException(e.getMessage(), e);
        }
    }

    @Override
    public boolean supports(final Class<?> authentication) {
        return authentication.equals(PreAuthenticatedAuthenticationToken.class);
    }

}

File: application/src/main/kotlin/br/com/jiratorio/security/TokenAuthenticationProvider.kt
package br.com.jiratorio.security

import br.com.jiratorio.service.AuthService
import org.springframework.security.authentication.AuthenticationProvider
import org.springframework.security.core.Authentication
import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken
import org.springframework.security.web.authentication.preauth.PreAuthenticatedCredentialsNotFoundException
import org.springframework.stereotype.Component

@Component
class TokenAuthenticationProvider(
    private val decodeToken: AuthService,
) : AuthenticationProvider {

    override fun authenticate(authentication: Authentication): Authentication {
        try {
            val account = decodeToken.decode(authentication.principal as String)
            return PreAuthenticatedAuthenticationToken(account, account.password, account.authorities)
        } catch (e: Exception) {
            throw PreAuthenticatedCredentialsNotFoundException(e.message, e)
        }
    }

    override fun supports(authentication: Class<*>): Boolean =
        authentication == PreAuthenticatedAuthenticationToken::class.java

}
